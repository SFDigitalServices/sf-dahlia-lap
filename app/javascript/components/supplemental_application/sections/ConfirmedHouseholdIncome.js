import React, { useEffect, useState } from 'react'
import { isNil, isString } from 'lodash'

import FormGrid from '~/components/molecules/FormGrid'
import { formatPercent } from '~/utils/utils'

import { YesNoRadioGroup, CurrencyField, PercentField, SelectField } from '~/utils/form/final_form/Field.js'
import validate from '~/utils/form/validations'

const validateIncomeCurrency = (value) => {
  return (
    validate.isValidCurrency('Please enter a valid dollar amount.')(value) ||
    validate.isUnderMaxValue(Math.pow(10, 15))('Please enter a smaller number.')(value)
  )
}

export const getAmiPercent = ({income, ami}) => {
  if (!income) {
    return 'Enter HH Income'
  }
  if (isNil(ami)) {
    return 'Missing AMI Chart'
  }
  let incomeFloat = isString(income) ? Number(income.replace(/[$,]+/g, '')) : income
  if (Number.isNaN(incomeFloat)) {
    return 'Fix HH Income'
  }
  return formatPercent(incomeFloat / Number(ami))
}

const ConfirmedHouseholdIncome = ({ listingAmiCharts, form, visited }) => {
  const { values } = form.getState()
  const currentAmiChartType = `${values.ami_chart_type} - ${values.ami_chart_year}`

  const [ amiChartTypes, setAmiChartTypes ] = useState([])
  const [ selectedType, setSelectedType ] = useState(currentAmiChartType || '')

  useEffect(() => {
    const getAmiCharts = async () => {
      if (listingAmiCharts.length > 0) {
        let chartTypes = listingAmiCharts.map((chart) => {
          const value = `${chart.ami_chart_type} - ${chart.ami_chart_year}`
          return { value, label: value }
        })

        if (chartTypes.length > 1) {
          chartTypes.unshift({ value: null, label: 'Select One...' })
        } else if (chartTypes.length === 1) {
          const initialType = chartTypes[0]
          setSelectedType(initialType)
          setFormYearAndType(initialType.value)
        }
        setAmiChartTypes(chartTypes)
      }
    }

    getAmiCharts()
  }, [])

  const setFormYearAndType = (combinedValue) => {
    const idx = combinedValue.lastIndexOf(' - ')
    const type = combinedValue.substr(0, idx)
    const year = combinedValue.substr(idx).replace(' - ', '')
    form.change('ami_chart_type', type)
    form.change('ami_chart_year', year)
  }

  const handleAmiSelectChange = ({ target: { value } }) => {
    setFormYearAndType(value)
    setSelectedType(value)
  }

  return (
    <React.Fragment>
      <FormGrid.Row>
        <FormGrid.Item>
          <FormGrid.Group label='Recurring Voucher/Subsidy'>
            <YesNoRadioGroup fieldName='housing_voucher_or_subsidy' />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group>
            <CurrencyField
              fieldName='household_assets'
              label='Household Assets'
              placeholder='Enter Amount'
              validation={validateIncomeCurrency}
              isDirty={visited && visited['household_assets']}
            />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group>
            <CurrencyField
              fieldName='confirmed_household_annual_income'
              label='Confirmed Annual Income'
              placeholder='Enter Amount'
              validation={validateIncomeCurrency}
              isDirty={visited && visited['confirmed_household_annual_income']}
            />
            <span className='form-note shift-up' id='household-annual-income'>
              Not Including % of Assets
            </span>
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group>
            <CurrencyField
              fieldName='hh_total_income_with_assets_annual'
              label='Final Household Annual Income'
              placeholder='Enter Amount'
              validation={validateIncomeCurrency}
              isDirty={visited && visited['hh_total_income_with_assets_annual']}
            />
            <span className='form-note shift-up' id='final-household-annual-income'>Includes % of assets if applicable</span>
          </FormGrid.Group>
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row paddingBottom>
        <FormGrid.Item>
          <PercentField
            id='ami_percentage'
            label='AMI Percentage'
            fieldName='ami_percentage'
            placeholder='Enter Percentage'
            validation={validate.isValidPercent('Please enter a valid percent.')}
            isDirty={visited && visited['ami_percentage']} />
        </FormGrid.Item>
        <FormGrid.Item>
          <SelectField
            id='ami_chart_type'
            fieldName='ami_chart_type'
            label='AMI Chart Type'
            onChange={handleAmiSelectChange}
            selectValue={selectedType}
            options={amiChartTypes}
            noPlaceholder />
        </FormGrid.Item>
      </FormGrid.Row>
    </React.Fragment>
  )
}

export default ConfirmedHouseholdIncome
