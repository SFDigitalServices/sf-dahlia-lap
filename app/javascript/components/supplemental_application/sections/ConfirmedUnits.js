import React from 'react'

import FormGrid from '~/components/molecules/FormGrid'
import { YesNoRadioGroup, CheckboxField } from '~/utils/form/final_form/Field.js'

const ConfirmedUnits = () => {
  return (
    <>
      <FormGrid.Row paddingBottom>
        <FormGrid.Item>
          <FormGrid.Group label='Senior in Household'>
            <YesNoRadioGroup fieldName='reserved_senior' trueValue='Yes' falseValue='No' />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group label='Veteran in Household'>
            <YesNoRadioGroup fieldName='has_military_service' trueValue='Yes' falseValue='No' />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group label='Person with Developmental Disability in Household'>
            <YesNoRadioGroup fieldName='has_developmental_disability' trueValue='Yes' falseValue='No' />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group label='Household ADA Priorities'>
            <div className='checkbox-group form-checkbox-item' role='group'>
              <CheckboxField
                fieldName='has_ada_priorities_selected.mobility_impairments'
                label='Mobility Impairments' />
              <CheckboxField
                fieldName='has_ada_priorities_selected.vision_impairments'
                label='Vision Impairments' />
              <CheckboxField
                fieldName='has_ada_priorities_selected.hearing_impairments'
                label='Hearing Impairments' />
            </div>
          </FormGrid.Group>
        </FormGrid.Item>
      </FormGrid.Row>
    </>
  )
}

export default ConfirmedUnits
