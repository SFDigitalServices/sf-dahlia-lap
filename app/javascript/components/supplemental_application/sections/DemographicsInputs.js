import React from 'react'

import { range } from 'lodash'

import FormGrid from 'components/molecules/FormGrid'
import { SelectField } from 'utils/form/final_form/Field'
import formUtils from 'utils/formUtils'

const DemographicsInputs = () => {
  const numberOfPeopleOptions = range(10)
  const martialStatusOptions = ['Single', 'Married', 'Domestic Partner']

  return (
    <>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            id='demographics-dependents'
            fieldName='number_of_dependents'
            options={numberOfPeopleOptions}
            format={formUtils.formatNumber}
            label='Number of Dependents'
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            id='demographics-seniors'
            fieldName='number_of_seniors'
            options={numberOfPeopleOptions}
            format={formUtils.formatNumber}
            label='Number of Seniors'
            helpText='Check the listing to verify the age cutoff for seniors.'
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            id='demographics-minors'
            fieldName='number_of_minors'
            options={numberOfPeopleOptions}
            format={formUtils.formatNumber}
            label='Number of Minors'
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            id='demographics-marital-status'
            fieldName='applicant.marital_status'
            options={martialStatusOptions}
            label='Primary Applicant Marital Status'
          />
        </FormGrid.Item>
      </FormGrid.Row>
    </>
  )
}

export default DemographicsInputs
