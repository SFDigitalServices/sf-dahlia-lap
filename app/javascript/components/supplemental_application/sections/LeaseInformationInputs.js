import React from 'react'
import { filter, map, isEmpty } from 'lodash'

import FormGrid from '~/components/molecules/FormGrid'
import { pluck } from '~/utils/utils'
import formUtils from '~/utils/formUtils'
import { withContext } from '../context'
import { SelectField, CurrencyField } from '~/utils/form/final_form/Field.js'
import { MultiDateField } from '~/utils/form/final_form/MultiDateField'
import { validateLeaseCurrency } from '~/utils/form/validations'

const toggleNoPreferenceUsed = (form, event) => {
  // lease.preference_used need to be reset, otherwise SF validation fails
  if (isEmpty(event.target.value)) {
    form.change('lease.preference_used', '')
  }
  form.change('lease.no_preference_used', isEmpty(event.target.value))
}

const LeaseInformationInputs = ({ form, store, visited }) => {
  const { availableUnits, application } = store
  const availableUnitsOptions = formUtils.toOptions(map(availableUnits, pluck('id', 'unit_number')))
  const noUnitsOptions = [{ value: '', label: 'No Units Available' }]
  const confirmedPreferences = filter(application.preferences, { 'post_lottery_validation': 'Confirmed' })
  const confirmedPreferenceOptions = formUtils.toOptions(map([{ 'id': null, 'preference_name': 'None' }, ...confirmedPreferences], pluck('id', 'preference_name')))
  return (
    <>
      <FormGrid.Row paddingBottom>
        <FormGrid.Item>
          <FormGrid.Group>
            <SelectField
              id='lease_assigned_unit'
              label='Assigned Unit Number'
              fieldName='lease.unit'
              options={availableUnitsOptions}
              disabled={!availableUnitsOptions.length}
              disabledOptions={noUnitsOptions} />
          </FormGrid.Group>
        </FormGrid.Item>
        <FormGrid.Item>
          <MultiDateField
            id='lease_start_date'
            fieldName='lease.lease_start_date'
            form={form}
            label='Lease Start Date' />
        </FormGrid.Item>
        <FormGrid.Item>
          <FormGrid.Group>
            <SelectField
              label='Preference Used'
              onChange={(value) => toggleNoPreferenceUsed(form, value)}
              fieldName='lease.preference_used'
              options={confirmedPreferenceOptions} />
          </FormGrid.Group>
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row paddingBottom>
        <FormGrid.Item>
          <CurrencyField
            label='Monthly Rent'
            fieldName='lease.total_monthly_rent_without_parking'
            placeholder='Enter Amount'
            validation={validateLeaseCurrency}
            isDirty={visited && visited['lease.total_monthly_rent_without_parking']} />
        </FormGrid.Item>
        <FormGrid.Item>
          <CurrencyField
            label='Monthly Tenant Contribution'
            fieldName='lease.monthly_tenant_contribution'
            placeholder='Enter Amount'
            validation={validateLeaseCurrency}
            isDirty={visited && visited['lease.monthly_tenant_contribution']} />
        </FormGrid.Item>
      </FormGrid.Row>
    </>
  )
}

export default withContext(LeaseInformationInputs)
