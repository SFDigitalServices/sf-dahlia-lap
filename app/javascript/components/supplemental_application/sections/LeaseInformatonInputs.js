import React from 'react'
import { Select, Text } from 'react-form'
import { map, toSafeInteger } from 'lodash'

import FormGrid from '~/components/molecules/FormGrid'
import { pluck, decorateComponents } from '~/utils/utils'
import { withNestedForm } from '~/utils/form/hoc'
import formUtils from '~/utils/formUtils'
import { withContext } from '../context'

const Plain = ({ text }) => <div className='text-value'>{text}</div>

const INPUTS = { 'Select': Select, 'Text': Text, 'Plain': Plain }

const CustomFormGrid = decorateComponents(INPUTS, Component => {
  return ({ label, ...rest }) => (
    <FormGrid.Item>
      <FormGrid.Group label={label}>
        <Component {...rest} />
      </FormGrid.Group>
    </FormGrid.Item>
  )
})

const getTotalMonthlyRent = (values) => {
  const monthlyUnitRent = values.total_monthly_rent_without_parking
  const monthlyParkingRent = values.monthly_parking_rent

  if (monthlyUnitRent || monthlyParkingRent) {
    return toSafeInteger(monthlyUnitRent) + toSafeInteger(monthlyParkingRent)
  } else {
    return 'Total Unit & Parking Rent'
  }
}

const LeaseInformationInputs = ({ formApi, store }) => {
  const { availableUnits } = store
  const availableUnitsOptions = formUtils.toOptions(map(availableUnits, pluck('id', 'unit_number')))
  const totalMonthlyRent = getTotalMonthlyRent(formApi.values)
  return (
    <React.Fragment>
      <FormGrid.Row paddingBottom>
        <CustomFormGrid.Select
          label='Assigned Unit Number'
          field='unit' options={availableUnitsOptions} placeholder='Select One' />
        <CustomFormGrid.Text
          field='lease_start_date'
          label='Lease Start Date'
          type='date'
          pattern='[0-9]{4}-[0-9]{2}-[0-9]{2}'
          placeholder='YYYY-MM-DD'
        />
        <CustomFormGrid.Select
          label='Preference Used'
          field='preference_used' options={[]} placeholder='Select One' />
      </FormGrid.Row>
      <FormGrid.Row paddingBottom>
        <CustomFormGrid.Text
          label='Monthly Unit Rent'
          field='total_monthly_rent_without_parking' type='number' placeholder='Enter Amount' />
        <CustomFormGrid.Text
          label='Monthly Parking Rent'
          field='monthly_parking_rent' type='number' placeholder='Enter Amount' />
        <CustomFormGrid.Plain
          label='Total Monthly Rent'
          text={String(totalMonthlyRent)} />
        <CustomFormGrid.Text
          label='Monthly Tenant Contribution'
          field='monthly_tenant_contribution' type='number' placeholder='Enter Amount' />
      </FormGrid.Row>
    </React.Fragment>
  )
}

export default withContext(withNestedForm('lease', LeaseInformationInputs))
