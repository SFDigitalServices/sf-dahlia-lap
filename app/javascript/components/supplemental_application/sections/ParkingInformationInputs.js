import React from 'react'

import FormGrid from '~/components/molecules/FormGrid'
import { CurrencyField, SelectField, InputField } from '~/utils/form/final_form/Field.js'
import { validateLeaseCurrency } from '~/utils/form/validations'

const monthlyRentFieldName = 'lease.monthly_parking_rent'
const parkingSpotFieldName = 'lease.parking_spot_number'
const Yes = 'Yes'
const No = 'No'
const Waitlist = 'Waitlist'

const ParkingInformationInputs = ({
  form: { change, getFieldState },
  values: { lease },
  disabled = false
}) => {
  const onChangeHasParkingSpace = ({ target: { value } }) => {
    if (value !== Yes) {
      change(monthlyRentFieldName, null)
      change(parkingSpotFieldName, null)
    }
  }

  const monthlyRentVisited = getFieldState(monthlyRentFieldName)?.visited
  const hasParkingSpace = lease?.bmr_parking_space_assigned === Yes

  return (
    <>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            label='BMR Parking Space Assigned?'
            onChange={onChangeHasParkingSpace}
            fieldName='lease.bmr_parking_space_assigned'
            options={[Yes, No, Waitlist]}
            disabled={disabled}
          />
        </FormGrid.Item>
        {hasParkingSpace && (
          <FormGrid.Item width='25%'>
            <CurrencyField
              label='Monthly Cost'
              fieldName={monthlyRentFieldName}
              validation={validateLeaseCurrency}
              disabled={disabled}
              isDirty={monthlyRentVisited}
              placeholder='Enter cost...'
            />
          </FormGrid.Item>
        )}
        {hasParkingSpace && (
          <FormGrid.Item width='25%'>
            <InputField
              label='Space Assigned'
              fieldName={parkingSpotFieldName}
              disabled={disabled}
              isDirty={monthlyRentVisited}
              placeholder='Enter space...'
            />
          </FormGrid.Item>
        )}
      </FormGrid.Row>
    </>
  )
}

export default ParkingInformationInputs
