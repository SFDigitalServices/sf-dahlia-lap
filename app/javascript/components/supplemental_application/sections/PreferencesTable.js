import React from 'react'
import { map, reject, overSome, findIndex, orderBy, kebabCase } from 'lodash'

import TableWrapper from '~/components/atoms/TableWrapper'
import PreferenceIcon from './preferences/PreferenceIcon'
import ExpandableTable from '~/components/molecules/ExpandableTable'
import Panel from './preferences/Panel'
import { isCOP, isDTHP, isAliceGriffith, getPreferenceName } from './preferences/utils'
import { getTypeOfProof } from './preferences/typeOfProof'
import { withContext } from '../context'
import { memberNameFromPref } from '~/components/applications/application_form/preferences/utils'

const { ExpanderButton } = ExpandableTable

const hasExpanderButton = (prefName) => !overSome(isCOP, isDTHP, isAliceGriffith)(prefName)

const onlyValid = (preferences) => {
  return reject(preferences, (pref) => {
    return !pref.receives_preference
  })
}

const matchingPreference = (row) => (preference) => {
  return getPreferenceName(preference) === row[1].content
}

/** Presenter **/

const buildRow = (proofFiles, applicationMembers, fileBaseUrl) => (preference) => {
  return [
    { content: <PreferenceIcon status={preference.post_lottery_validation} /> },
    { content: getPreferenceName(preference) },
    { content: memberNameFromPref(preference.application_member_id, applicationMembers) },
    { content: preference.preference_lottery_rank, classes: ['text-right'] },
    { content: getTypeOfProof(preference, proofFiles, fileBaseUrl) },
    { content: preference.post_lottery_validation }
  ]
}

const buildRows = (application, applicationMembers, fileBaseUrl) => {
  const { preferences } = application
  const proofFiles = application.proof_files
  const sortedPreferences = orderBy(preferences, 'preference_order', 'asc')
  return map(onlyValid(sortedPreferences), buildRow(proofFiles, applicationMembers, fileBaseUrl))
}

const columns = [
  { content: '' },
  { content: 'Preference' },
  { content: 'Claimant' },
  { content: 'Rank', classes: ['text-right'] },
  { content: 'Proof' },
  { content: 'Status' },
  { content: '' }
]

const expandedRowRenderer = (
  application,
  applicationMembers,
  onSave,
  onPanelClose,
  form,
  visited
) => (row, toggle) => {
  const preferenceIndex = findIndex(application.preferences, matchingPreference(row))
  const handleOnClose = (preferenceIndex) => {
    toggle()
    onPanelClose && onPanelClose(preferenceIndex)
  }
  const handleOnSave = async (preferenceIndex, application) => {
    const response = await onSave(preferenceIndex, application)
    response && handleOnClose()
  }

  return (
    <Panel
      application={application}
      applicationMembers={applicationMembers}
      preferenceIndex={preferenceIndex}
      onSave={handleOnSave}
      onClose={handleOnClose}
      form={form}
      visited={visited}
    />
  )
}

const expanderAction = (row, expanded, expandedRowToggler) => {
  const prefName = row[1].content
  return (
    !expanded &&
    hasExpanderButton(prefName) && (
      <ExpanderButton
        label='Edit'
        onClick={expandedRowToggler}
        id={`${kebabCase(prefName)}-edit`}
      />
    )
  )
}

const PreferencesTable = ({
  application,
  applicationMembers,
  fileBaseUrl,
  onSave,
  onPanelClose,
  form,
  visited
}) => {
  const rows = buildRows(application, applicationMembers, fileBaseUrl)
  return (
    <div className='preferences-table'>
      <TableWrapper>
        <ExpandableTable
          columns={columns}
          rows={rows}
          rowKeyIndex={1}
          expanderRenderer={expanderAction}
          expandedRowRenderer={expandedRowRenderer(
            application,
            applicationMembers,
            onSave,
            onPanelClose,
            form,
            visited
          )}
        />
      </TableWrapper>
    </div>
  )
}

export default withContext(PreferencesTable)
