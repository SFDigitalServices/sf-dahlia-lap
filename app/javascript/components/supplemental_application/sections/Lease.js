import React, { useState } from 'react'
import { filter, map, isEmpty } from 'lodash'

import Button from '~/components/atoms/Button'
import ContentSection from '~/components/molecules/ContentSection'
import FormGrid from '~/components/molecules/FormGrid'
import InlineModal from '~/components/molecules/InlineModal'
import formUtils from '~/utils/formUtils'
import ParkingInformationInputs from './ParkingInformationInputs'
import RentalAssistance from './RentalAssistance'
import { convertPercentAndCurrency } from '../../../utils/form/validations'
import ConfirmationModal from '~/components/organisms/ConfirmationModal'

import { pluck } from '~/utils/utils'
import { withContext } from '../context'
import { CurrencyField, Label, SelectField } from '~/utils/form/final_form/Field.js'
import { MultiDateField } from '~/utils/form/final_form/MultiDateField'
import { validateLeaseCurrency } from '~/utils/form/validations'
import { EDIT_LEASE_STATE } from '../SupplementalApplicationPage'
import { doesApplicationHaveLease } from '~/utils/leaseUtils'
import { areLeaseAndRentalAssistancesValid } from '~/utils/form/formSectionValidations'

const toggleNoPreferenceUsed = (form, event) => {
  // lease.preference_used need to be reset, otherwise SF validation fails
  if (isEmpty(event.target.value)) {
    form.change('lease.preference_used', '')
  }
  form.change('lease.no_preference_used', isEmpty(event.target.value))
}
const LeaseActions = ({
  onEditLeaseClick,
  onSave,
  onCancelLeaseClick,
  onDelete,
  leaseExists,
  loading,
  isEditing
}) => {
  if (!isEditing) {
    return (
      <div className='form-grid_item column'>
        <Button
          id='edit-lease-button'
          classes='secondary'
          small
          onClick={onEditLeaseClick}
          disabled={loading}
          noBottomMargin
          text='Edit Lease'
        />
      </div>
    )
  }

  return (
    <div className='form-grid_item column'>
      <Button
        classes='primary margin-right'
        small
        onClick={onSave}
        disabled={loading}
        noBottomMargin
        text='Save Lease'
      />
      <Button
        classes='secondary'
        small
        onClick={onCancelLeaseClick}
        disabled={loading}
        noBottomMargin
        text='Cancel'
      />
      {leaseExists && (
        <Button
          classes='alert-fill right'
          small
          onClick={onDelete}
          disabled={loading}
          noBottomMargin
          text='Delete'
        />
      )}
    </div>
  )
}

const Lease = ({ form, values, store }) => {
  const {
    availableUnits,
    application,
    handleSaveLease,
    handleDeleteLease,
    handleEditLeaseClick,
    handleCancelLeaseClick,
    leaseSectionState,
    loading
  } = store

  const [showDeleteConfirmation, setShowDeleteConfirmation] = useState(false)

  const isEditingMode = leaseSectionState === EDIT_LEASE_STATE
  const disabled = !isEditingMode

  const availableUnitsOptions = formUtils.toOptions(
    map(availableUnits, pluck('id', 'unit_number'))
  )

  const noUnitsOptions = [{ value: '', label: 'No Units Available' }]
  const confirmedPreferences = filter(application.preferences, {
    post_lottery_validation: 'Confirmed'
  })

  const confirmedPreferenceOptions = formUtils.toOptions(
    map(
      [{ id: null, preference_name: 'None' }, ...confirmedPreferences],
      pluck('id', 'preference_name')
    )
  )

  const getVisited = (fieldName) => (
    form.getFieldState(fieldName)?.visited
  )

  const openDeleteLeaseConfirmation = () => setShowDeleteConfirmation(true)
  const closeDeleteLeaseConfirmation = () => setShowDeleteConfirmation(false)

  const handleDeleteLeaseConfirmed = () => {
    setShowDeleteConfirmation(false)
    handleDeleteLease()
  }

  const areNoUnitsAvailable = !availableUnitsOptions.length

  const validateAndSaveLease = (form) => {
    if (areLeaseAndRentalAssistancesValid(form)) {
      handleSaveLease(convertPercentAndCurrency(form.getState().values))
    } else {
      // submit to force errors to display
      form.submit()
    }
  }

  return (
    <InlineModal>
      <ContentSection.Header description='If the household receives recurring rental assistance, remember to subtract this from the unitâ€™s rent when calculating Tenant Contribution.' />
      <ContentSection.Sub title='Unit and Parking'>
        <FormGrid.Row>
          <FormGrid.Item>
            <SelectField
              id='lease_assigned_unit'
              label='Assigned Unit Number'
              fieldName='lease.unit'
              options={availableUnitsOptions}
              disabled={disabled || areNoUnitsAvailable}
              disabledOptions={areNoUnitsAvailable && noUnitsOptions}
            />
          </FormGrid.Item>
        </FormGrid.Row>
        <ParkingInformationInputs
          form={form}
          values={values}
          disabled={disabled}
        />
      </ContentSection.Sub>
      <ContentSection.Sub
        title='Rent and Assistance'
      >
        <FormGrid.Row>
          <FormGrid.Item>
            <CurrencyField
              label='Monthly Rent'
              fieldName='lease.total_monthly_rent_without_parking'
              validation={validateLeaseCurrency}
              isDirty={
                getVisited('lease.total_monthly_rent_without_parking')
              }
              disabled={disabled}
            />
          </FormGrid.Item>
        </FormGrid.Row>
        <FormGrid.Row>
          <FormGrid.Item>
            <Label label='Rental Assistance' />
          </FormGrid.Item>
        </FormGrid.Row>
        <RentalAssistance
          form={form}
          disabled={disabled}
          loading={loading}
        />
        <FormGrid.Row>
          <FormGrid.Item>
            <CurrencyField
              label='Tenant Contribution'
              fieldName='lease.monthly_tenant_contribution'
              helpText='Monthly rent minus recurring rental assistance, if any'
              validation={validateLeaseCurrency}
              isDirty={getVisited('lease.monthly_tenant_contribution')}
              disabled={disabled}
            />
          </FormGrid.Item>
        </FormGrid.Row>
      </ContentSection.Sub>
      <ContentSection.Sub title='Lease Details'>
        <FormGrid.Row>
          <FormGrid.Item>
            <MultiDateField
              id='lease_start_date'
              fieldName='lease.lease_start_date'
              form={form}
              label='Lease Start Date'
              disabled={disabled}
            />
          </FormGrid.Item>
        </FormGrid.Row>
        <FormGrid.Row>
          <FormGrid.Item>
            <SelectField
              label='Preference Used'
              onChange={value => toggleNoPreferenceUsed(form, value)}
              fieldName='lease.preference_used'
              options={confirmedPreferenceOptions}
              disabled={disabled}
            />
          </FormGrid.Item>
        </FormGrid.Row>
      </ContentSection.Sub>
      <FormGrid.Row>
        <LeaseActions
          onSave={() => validateAndSaveLease(form)}
          onCancelLeaseClick={() => handleCancelLeaseClick(form)}
          onEditLeaseClick={handleEditLeaseClick}
          onDelete={openDeleteLeaseConfirmation}
          loading={loading}
          leaseExists={doesApplicationHaveLease(application)}
          isEditing={isEditingMode}
        />
      </FormGrid.Row>
      <ConfirmationModal
        isOpen={showDeleteConfirmation}
        onCloseClick={closeDeleteLeaseConfirmation}
        onSecondaryClick={closeDeleteLeaseConfirmation}
        onPrimaryClick={handleDeleteLeaseConfirmed}
        primaryButtonIsAlert
        primaryText='Delete Lease'
        secondaryText='Continue Editing'
        subtitle='This will permanently delete the lease and all related rental assistances.'
        title='Are you sure you want to delete this lease?'
      />
    </InlineModal>
  )
}

export default withContext(Lease)
