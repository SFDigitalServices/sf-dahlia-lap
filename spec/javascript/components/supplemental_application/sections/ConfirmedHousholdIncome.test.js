import React from 'react'
import { getAmiPercent } from 'components/supplemental_application/sections/ConfirmedHouseholdIncome'
import ConfirmedHouseholdIncome from '~/components/supplemental_application/sections/ConfirmedHouseholdIncome'
import { withForm } from '../../../testUtils/wrapperUtil'

jest.mock('apiService', () => ({
  getAMI: async (data) => {
    return {
      ami: [
        { chartType: data.chartType, year: data.chartYear, amount: 100, numOfHousehold: 1 },
        { chartType: data.chartType, year: data.chartYear, amount: 200, numOfHousehold: 2 },
        { chartType: data.chartType, year: data.chartYear, amount: 300, numOfHousehold: 3 }
      ]
    }
  }
}))

const getWrapper = (application, mockAmiCharts, shouldMount = false) => {
  return withForm(
    application,
    (form) => <ConfirmedHouseholdIncome visited={{}} listingAmiCharts={mockAmiCharts} />,
    shouldMount
  )
}

describe('ConfirmedHouseholdIncome', () => {
  describe('getAmiPercent', () => {
    test('it returns percentage of ami for given income', () => {
      expect(getAmiPercent({ income: 10000, ami: '100000' })).toBe('10%')
      expect(getAmiPercent({ income: '10000', ami: '100000' })).toBe('10%')
      expect(getAmiPercent({ income: '$10,000', ami: '100000' })).toBe('10%')
      expect(getAmiPercent({ income: '$10,000.00', ami: '100000' })).toBe('10%')
    })
    test('it returns useful messages if params are missing', () => {
      expect(getAmiPercent({ income: null, ami: '100000' })).toBe('Enter HH Income')
      expect(getAmiPercent({ income: '$1,000', ami: null })).toBe('Missing AMI Chart')
    })
    test('it returns error message if income is invalid', () => {
      expect(getAmiPercent({ income: 'zzz', ami: '100000' })).toBe('Fix HH Income')
    })
  })

  test('it matches snapshot when all values are empty', () => {
    const wrapper = getWrapper({}, [])
    expect(wrapper).toMatchSnapshot()
  })

  describe('getAmiCharts', () => {
    const mockAmiCharts = [
      {
        ami_chart_type: 'chart1',
        ami_chart_year: 2020
      },
      {
        ami_chart_type: 'chart2',
        ami_chart_year: 2020
      }
    ]
    test('it sets chart options correctly when charts are empty', () => {
      const wrapper = getWrapper({}, [], true)
      expect(wrapper.find('SelectField#ami_chart_type').props().options).toEqual([])
      expect(wrapper.find('SelectField#ami_chart_year').props().options).toEqual([])
    })
    test('it presents all available chart types when provided', () => {
      const wrapper = getWrapper({}, mockAmiCharts, true)
      const expectedChartOptions = [
        {
          label: 'chart1',
          value: 'chart1'
        },
        {
          label: 'chart2',
          value: 'chart2'
        }
      ]
      const expectedYearOptions = [{ label: 2020, value: 2020 }]
      expect(wrapper.find('SelectField#ami_chart_type').props().options).toEqual(
        expectedChartOptions
      )
      expect(wrapper.find('SelectField#ami_chart_year').props().options).toEqual(
        expectedYearOptions
      )
    })

    test('it fills dates to present year when AMI chart year is in the past', () => {
      const mockOldAmiCharts = [{ ami_chart_type: 'chart', ami_chart_year: 2018 }]
      const wrapper = getWrapper({}, mockOldAmiCharts, true)
      const expectedYearOptions = [
        { label: 2018, value: 2018 },
        { label: 2019, value: 2019 },
        { label: 2020, value: 2020 }
      ]
      expect(wrapper.find('SelectField#ami_chart_year').props().options).toEqual(
        expectedYearOptions
      )
    })
  })
})
