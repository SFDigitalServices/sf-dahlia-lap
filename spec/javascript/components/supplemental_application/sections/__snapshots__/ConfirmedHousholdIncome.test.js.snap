// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmedHouseholdIncome it matches snapshot when all values are empty 1`] = `
<Fragment>
  <Component
    title="Confirmed Household Income"
  >
    <Component
      expand={true}
    >
      <Component>
        <Component
          label="Household claimed a recurring voucher or subsidy"
        >
          <YesNoRadioGroup
            fieldName="housing_voucher_or_subsidy"
          />
        </Component>
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <CurrencyField
          fieldName="imputed_income_from_assets"
          label="Imputed Income from Assets"
          validation={[Function]}
        />
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <CurrencyField
          fieldName="household_assets"
          label="Total Household Liquid Assets"
          validation={[Function]}
        />
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <CurrencyField
          fieldName="confirmed_household_annual_income"
          helpText="Not including imputed income from assets"
          label="Total Household Annual Income"
          validation={[Function]}
        />
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <CurrencyField
          fieldName="hh_total_income_with_assets_annual"
          helpText="Includes imputed income from assets, if applicable"
          label="Final Household Annual Income"
          validation={[Function]}
        />
      </Component>
    </Component>
  </Component>
  <Component
    title="Area Median Income"
  >
    <Component
      expand={true}
    >
      <Component>
        <PercentField
          fieldName="ami_percentage"
          id="ami_percentage"
          label="Household AMI Percentage"
          validation={[Function]}
        />
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <SelectField
          fieldName="ami_chart_type"
          id="ami_chart_type"
          label="AMI Chart Type"
          noPlaceholder={true}
          options={Array []}
        />
      </Component>
    </Component>
    <Component
      expand={true}
    >
      <Component>
        <SelectField
          fieldName="ami_chart_year"
          id="ami_chart_year"
          label="AMI Chart Year"
          noPlaceholder={true}
          options={Array []}
        />
      </Component>
    </Component>
  </Component>
</Fragment>
`;
